src/math/barycenter.cpp

```diff
@@ -41,35 +41,58 @@ Eigen::Vector2d calcBarycentricCoordsForEdge(
   return barycentricCoords;
 }
 
+static double crossProduct2D(const Eigen::Vector2d &u, const Eigen::Vector2d &v)
+{
+  return u(0) * v(1) - u(1) * v(0);
+}
+
 Eigen::Vector3d calcBarycentricCoordsForTriangle(
     const Eigen::VectorXd &a,
     const Eigen::VectorXd &b,
     const Eigen::VectorXd &c,
     const Eigen::VectorXd &u)
 {
+  using Eigen::Vector2d;
   using Eigen::Vector3d;
 
   const int dimensions = a.size();
-  PRECICE_ASSERT(dimensions == 3, dimensions);
   PRECICE_ASSERT(dimensions == b.size(), "A and B need to have the same dimensions.", dimensions, b.size());
   PRECICE_ASSERT(dimensions == c.size(), "A and C need to have the same dimensions.", dimensions, c.size());
   PRECICE_ASSERT(dimensions == u.size(), "A and the point need to have the same dimensions.", dimensions, u.size());
 
-  Vector3d ab, ac, au, n, barycentricCoords;
+  Vector3d barycentricCoords;
   double   scaleFactor;
 
   // constant per triangle
-  ab          = b - a;
-  ac          = c - a;
-  n           = ab.cross(ac);
-  scaleFactor = 1.0 / n.dot(n);
-
-  // varying per point
-  au = u - a;
-
-  barycentricCoords(2) = n.dot(ab.cross(au)) * scaleFactor;
-  barycentricCoords(1) = n.dot(au.cross(ac)) * scaleFactor;
-  barycentricCoords(0) = 1 - barycentricCoords(1) - barycentricCoords(2);
+  if (dimensions == 3) {
+    Vector3d ab, ac, au, n;
+
+    ab          = b - a;
+    ac          = c - a;
+    n           = ab.cross(ac);
+    scaleFactor = 1.0 / n.dot(n);
+
+    // varying per point
+    au = u - a;
+
+    barycentricCoords(2) = n.dot(ab.cross(au)) * scaleFactor;
+    barycentricCoords(1) = n.dot(au.cross(ac)) * scaleFactor;
+    barycentricCoords(0) = 1 - barycentricCoords(1) - barycentricCoords(2);
+
+  } else {
+    // Dimension == 2 => No cross product
+    Vector2d ab, ac, ub, uc, ua;
+    ab = b - a;
+    ac = c - a;
+    ub = b - u;
+    uc = c - u;
+    ua = a - u;
+
+    scaleFactor          = 1.0 / crossProduct2D(ab, ac);
```

Comment:
This will crash for zero vectors. Can you assert this above?

Reply:
It wasn't checked in 3D either, so I just added assertions for both cases.